File: solutions/cpp/Rearrange.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::swap;
using std::uniform_int_distribution;
using std::vector;

// @include
void Rearrange(vector<int>* A) {
  auto& A_ref = *A;
  for (size_t i = 1; i < A_ref.size(); ++i) {
    if ((!(i & 1) && A_ref[i - 1] < A_ref[i]) ||
        ((i & 1) && A_ref[i - 1] > A_ref[i])) {
      swap(A_ref[i - 1], A_ref[i]);
    }
  }
}
// @exclude

void check_answer(const vector<int>& A) {
  for (size_t i = 0; i < A.size(); ++i) {
    if (i & 1) {
      assert(A[i] >= A[i - 1]);
      if (i + 1 < A.size()) {
        assert(A[i] >= A[i + 1]);
      }
    } else {
      if (i > 0) {
        assert(A[i - 1] >= A[i]);
      }
      if (i + 1 < A.size()) {
        assert(A[i + 1] >= A[i]);
      }
    }
  }
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 10000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<int> A;
    uniform_int_distribution<int> dis(-n, n);
    for (int i = 0; i < n; ++i) {
      A.emplace_back(dis(gen));
    }
    Rearrange(&A);
    /*
    for (const int &a : A) {
      cout << a << ' ';
    }
    cout << endl;
    */
    check_answer(A);
  }
  return 0;
}

File: solutions/python/rearrange.py
import sys
import random


# @include
def rearrange(A):
    for i in range(len(A)):
        A[i:i + 2] = sorted(A[i:i + 2], reverse=i % 2)
# @exclude


def check_answer(A):
    for i in range(len(A)):
        if i % 2:
            assert A[i] >= A[i - 1]
            if i + 1 < len(A):
                assert A[i] >= A[i + 1]
        else:
            if i > 0:
                assert A[i - 1] >= A[i]
            if i + 1 < len(A):
                assert A[i + 1] >= A[i]


def main():
    for _ in range(10000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10000)
        A = [random.randint(-n, n) for _ in range(n)]
        rearrange(A)
        check_answer(A)


if __name__ == '__main__':
    main()
